{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __ResFinder__\n",
    "\n",
    "Search carbapenemase gen occurrences using illumina reads directly\n",
    "\n",
    "__Dependences__\n",
    "\n",
    "- KMA (bioconda) -> (cd /KMA && make)\n",
    "- BLAST (bioconda)\n",
    "- resfinder (pip)\n",
    "\n",
    "Databases\n",
    "\n",
    "(in ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/)\n",
    "\n",
    "- git clone https://bitbucket.org/genomicepidemiology/resfinder_db/\n",
    "- git clone https://bitbucket.org/genomicepidemiology/pointfinder_db/\n",
    "- git clone https://bitbucket.org/genomicepidemiology/disinfinder_db/\n",
    "\n",
    "(luego en cada repositorio git clonado hay que ejecutar el INSTALL.py)\n",
    "\n",
    "Por alguna razon, pese a que resfinder es un programa en python, est√° construido solamente para ejecutarlo mediante CLI xD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: __main__.py [-h] [-ifa INPUTFASTA] [-ifq INPUTFASTQ [INPUTFASTQ ...]]\n",
      "                   [--nanopore] -o OUTPUTPATH [-j OUT_JSON] [-b BLASTPATH]\n",
      "                   [-k KMAPATH] [-s SPECIES] [--ignore_missing_species]\n",
      "                   [--output_aln] [-db_res DB_PATH_RES]\n",
      "                   [-db_res_kma DB_PATH_RES_KMA] [-acq] [-ao ACQ_OVERLAP]\n",
      "                   [-l MIN_COV] [-t THRESHOLD] [-d]\n",
      "                   [-db_disinf DB_PATH_DISINF]\n",
      "                   [-db_disinf_kma DB_PATH_DISINF_KMA] [-c]\n",
      "                   [-db_point DB_PATH_POINT] [-db_point_kma DB_PATH_POINT_KMA]\n",
      "                   [-g SPECIFIC_GENE [SPECIFIC_GENE ...]] [-u]\n",
      "                   [-l_p MIN_COV_POINT] [-t_p THRESHOLD_POINT]\n",
      "                   [--ignore_indels] [--ignore_stop_codons] [-v] [--pickle]\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  -ifa INPUTFASTA, --inputfasta INPUTFASTA\n",
      "                        Input fasta file.\n",
      "  -ifq INPUTFASTQ [INPUTFASTQ ...], --inputfastq INPUTFASTQ [INPUTFASTQ ...]\n",
      "                        Input fastq file(s). Assumed to be single-end fastq if\n",
      "                        only one file is provided, and assumed to be paired-\n",
      "                        end data if two files are provided.\n",
      "  --nanopore            If nanopore data is used\n",
      "  -o OUTPUTPATH, --outputPath OUTPUTPATH\n",
      "                        Output directory. If it doesn't exist, it will be\n",
      "                        created.\n",
      "  -j OUT_JSON, --out_json OUT_JSON\n",
      "                        Specify JSON filename and output directory. If the\n",
      "                        directory doesn't exist, it will be created.\n",
      "  -b BLASTPATH, --blastPath BLASTPATH\n",
      "                        Path to blastn\n",
      "  -k KMAPATH, --kmaPath KMAPATH\n",
      "                        Path to KMA\n",
      "  -s SPECIES, --species SPECIES\n",
      "                        Species in the sample\n",
      "  --ignore_missing_species\n",
      "                        If set, species is provided and --point flag is set,\n",
      "                        will not throw an error if no database is found for\n",
      "                        the provided species. If species is not found. Point\n",
      "                        mutations will silently be ignored.\n",
      "  --output_aln          will add the alignments in the json output.\n",
      "  -db_res DB_PATH_RES, --db_path_res DB_PATH_RES\n",
      "                        Path to the databases for ResFinder.\n",
      "  -db_res_kma DB_PATH_RES_KMA, --db_path_res_kma DB_PATH_RES_KMA\n",
      "                        Path to the ResFinder databases indexed with KMA.\n",
      "                        Defaults to the value of the --db_res flag.\n",
      "  -acq, --acquired      Run resfinder for acquired resistance genes\n",
      "  -ao ACQ_OVERLAP, --acq_overlap ACQ_OVERLAP\n",
      "                        Genes are allowed to overlap this number of\n",
      "                        nucleotides. Default: 30.\n",
      "  -l MIN_COV, --min_cov MIN_COV\n",
      "                        Minimum (breadth-of) coverage of ResFinder within the\n",
      "                        range 0-1.\n",
      "  -t THRESHOLD, --threshold THRESHOLD\n",
      "                        Threshold for identity of ResFinder within the range\n",
      "                        0-1.\n",
      "  -d, --disinfectant    Run resfinder for disinfectant resistance genes\n",
      "  -db_disinf DB_PATH_DISINF, --db_path_disinf DB_PATH_DISINF\n",
      "                        Path to the databases for DisinFinder.\n",
      "  -db_disinf_kma DB_PATH_DISINF_KMA, --db_path_disinf_kma DB_PATH_DISINF_KMA\n",
      "                        Path to the DisinFinder databases indexed with KMA.\n",
      "                        Defaults to the value of the --db_res flag.\n",
      "  -c, --point           Run pointfinder for chromosomal mutations\n",
      "  -db_point DB_PATH_POINT, --db_path_point DB_PATH_POINT\n",
      "                        Path to the databases for PointFinder\n",
      "  -db_point_kma DB_PATH_POINT_KMA, --db_path_point_kma DB_PATH_POINT_KMA\n",
      "                        Path to the PointFinder databases indexed with KMA.\n",
      "                        Defaults to the value of the --db_path_point flag.\n",
      "  -g SPECIFIC_GENE [SPECIFIC_GENE ...], --specific_gene SPECIFIC_GENE [SPECIFIC_GENE ...]\n",
      "                        Specify genes existing in the database to search for -\n",
      "                        if none is specified all genes are included in the\n",
      "                        search.\n",
      "  -u, --unknown_mut     Show all mutations found even if in unknown to the\n",
      "                        resistance database\n",
      "  -l_p MIN_COV_POINT, --min_cov_point MIN_COV_POINT\n",
      "                        Minimum (breadth-of) coverage of Pointfinder within\n",
      "                        the range 0-1. If None is selected, the minimum\n",
      "                        coverage of ResFinder will be used.\n",
      "  -t_p THRESHOLD_POINT, --threshold_point THRESHOLD_POINT\n",
      "                        Threshold for identity of Pointfinder within the range\n",
      "                        0-1. If None is selected, the minimum coverage of\n",
      "                        ResFinder will be used.\n",
      "  --ignore_indels       Ignore frameshift-causing indels in Pointfinder.\n",
      "  --ignore_stop_codons  Ignore premature stop codons in Pointfinder.\n",
      "  -v, --version         Show program's version number and exit\n",
      "  --pickle              Create a pickle dump of the Isolate object. Currently\n",
      "                        needed in the CGE webserver. Dependency and this\n",
      "                        option is being removed.\n"
     ]
    }
   ],
   "source": [
    "!python -m resfinder --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "samples=\"/srv/storage/oceania@storage2.lille.grid5000.fr/npilquinao/ISP/2carbapenemase/samples\"\n",
    "out=\"/home/npilquinao/BEM-ISP/ResFinder\"\n",
    "databases=\"/srv/storage/oceania@storage2.lille.grid5000.fr/npilquinao/ISP/2carbapenemase/resfinder_databases\"\n",
    "\n",
    "python -m resfinder\\\n",
    "    --inputfastq\\\n",
    "        $samples/VA61_2022/61_22_S82_R1_001.fastq.gz\\\n",
    "        $samples/VA61_2022/61_22_S82_R2_001.fastq.gz\\\n",
    "    --outputPath $out/VA61_2022/\\\n",
    "    --acquired --disinfectant\\\n",
    "    --db_path_res $databases/resfinder_db\\\n",
    "    --db_path_disinf $databases/disinfinder_db\\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Resfinder output__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Diferent resistance genes: 27\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (27, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Resistance gene</th><th>Identity</th><th>Alignment Length/Gene Length</th><th>Coverage</th><th>Position in reference</th><th>Contig</th><th>Position in contig</th><th>Phenotype</th><th>Accession no.</th><th>Sample</th></tr><tr><td>str</td><td>f64</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;aph(6)-Id&quot;</td><td>100.0</td><td>&quot;837/837&quot;</td><td>100.0</td><td>&quot;1..838&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Streptomycin&quot;</td><td>&quot;M28829&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;ant(3&#x27;&#x27;)-Ia&quot;</td><td>98.15</td><td>&quot;989/972&quot;</td><td>101.03</td><td>&quot;1..990&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Streptomycin&quot;</td><td>&quot;X02340&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;aac(6&#x27;)-Ib-cr&quot;</td><td>100.0</td><td>&quot;600/600&quot;</td><td>100.0</td><td>&quot;1..601&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Tobramycin, Di‚Ä¶</td><td>&quot;DQ303918&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;aph(3&#x27;&#x27;)-Ib&quot;</td><td>99.0</td><td>&quot;804/804&quot;</td><td>99.88</td><td>&quot;1..805&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Streptomycin&quot;</td><td>&quot;AF321551&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;aac(3)-IIa&quot;</td><td>100.0</td><td>&quot;861/861&quot;</td><td>100.0</td><td>&quot;1..862&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Apramycin, Gen‚Ä¶</td><td>&quot;CP023555&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;sul1&quot;</td><td>99.88</td><td>&quot;867/867&quot;</td><td>100.0</td><td>&quot;1..868&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Sulfamethoxazo‚Ä¶</td><td>&quot;EU780013&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;sul2&quot;</td><td>100.0</td><td>&quot;816/816&quot;</td><td>100.0</td><td>&quot;1..817&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Sulfamethoxazo‚Ä¶</td><td>&quot;AY034138&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;tet(A)&quot;</td><td>97.25</td><td>&quot;1240/1275&quot;</td><td>97.25</td><td>&quot;1..1241&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Doxycycline, T‚Ä¶</td><td>&quot;AF534183&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;dfrA14&quot;</td><td>83.02</td><td>&quot;483/483&quot;</td><td>83.44</td><td>&quot;1..484&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Trimethoprim&quot;</td><td>&quot;DQ388123&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;dfrA14&quot;</td><td>100.0</td><td>&quot;474/474&quot;</td><td>100.0</td><td>&quot;1..475&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Trimethoprim&quot;</td><td>&quot;KF921535&quot;</td><td>&quot;VA1046_2020&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (27, 10)\n",
       "‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê\n",
       "‚îÇ Resistance ‚îÜ Identity ‚îÜ Alignment ‚îÜ Coverage ‚îÜ ‚Ä¶ ‚îÜ Position  ‚îÜ Phenotype ‚îÜ Accession ‚îÜ Sample    ‚îÇ\n",
       "‚îÇ gene       ‚îÜ ---      ‚îÜ Length/Ge ‚îÜ ---      ‚îÜ   ‚îÜ in contig ‚îÜ ---       ‚îÜ no.       ‚îÜ ---       ‚îÇ\n",
       "‚îÇ ---        ‚îÜ f64      ‚îÜ ne Length ‚îÜ f64      ‚îÜ   ‚îÜ ---       ‚îÜ str       ‚îÜ ---       ‚îÜ str       ‚îÇ\n",
       "‚îÇ str        ‚îÜ          ‚îÜ ---       ‚îÜ          ‚îÜ   ‚îÜ str       ‚îÜ           ‚îÜ str       ‚îÜ           ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ str       ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ           ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°\n",
       "‚îÇ aph(6)-Id  ‚îÜ 100.0    ‚îÜ 837/837   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Streptomy ‚îÜ M28829    ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ ant(3'')-I ‚îÜ 98.15    ‚îÜ 989/972   ‚îÜ 101.03   ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Streptomy ‚îÜ X02340    ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ a          ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ aac(6')-Ib ‚îÜ 100.0    ‚îÜ 600/600   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Tobramyci ‚îÜ DQ303918  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ -cr        ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ n, Dibeka ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin, Amik ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ acin,‚Ä¶    ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ aph(3'')-I ‚îÜ 99.0     ‚îÜ 804/804   ‚îÜ 99.88    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Streptomy ‚îÜ AF321551  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ b          ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ aac(3)-IIa ‚îÜ 100.0    ‚îÜ 861/861   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Apramycin ‚îÜ CP023555  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ , Gentami ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin, Tobr ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ amyci‚Ä¶    ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ ‚Ä¶          ‚îÜ ‚Ä¶        ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶        ‚îÜ ‚Ä¶ ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶         ‚îÇ\n",
       "‚îÇ sul1       ‚îÜ 99.88    ‚îÜ 867/867   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Sulfameth ‚îÜ EU780013  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ oxazole   ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ sul2       ‚îÜ 100.0    ‚îÜ 816/816   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Sulfameth ‚îÜ AY034138  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ oxazole   ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ tet(A)     ‚îÜ 97.25    ‚îÜ 1240/1275 ‚îÜ 97.25    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Doxycycli ‚îÜ AF534183  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ ne, Tetra ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cycline   ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ dfrA14     ‚îÜ 83.02    ‚îÜ 483/483   ‚îÜ 83.44    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Trimethop ‚îÜ DQ388123  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ rim       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ dfrA14     ‚îÜ 100.0    ‚îÜ 474/474   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Trimethop ‚îÜ KF921535  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ rim       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['aph(6)-Id',\n",
       " \"ant(3'')-Ia\",\n",
       " \"aac(6')-Ib-cr\",\n",
       " \"aph(3'')-Ib\",\n",
       " 'aac(3)-IIa',\n",
       " 'blaOXA-1',\n",
       " 'blaOXA-10',\n",
       " 'blaSHV-2a',\n",
       " 'blaSHV-27',\n",
       " 'blaSHV-110',\n",
       " 'blaTEM-1B',\n",
       " 'blaNDM-7',\n",
       " 'blaCTX-M-2',\n",
       " 'fosA',\n",
       " 'fosA',\n",
       " 'cmlA1',\n",
       " 'qnrB1',\n",
       " 'qnrB19',\n",
       " 'OqxA',\n",
       " 'OqxB',\n",
       " 'ARR-3',\n",
       " 'sul1',\n",
       " 'sul1',\n",
       " 'sul2',\n",
       " 'tet(A)',\n",
       " 'dfrA14',\n",
       " 'dfrA14']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['Streptomycin',\n",
       " 'Streptomycin',\n",
       " 'Tobramycin, Dibekacin, Amikacin, Sisomicin, Netilmicin, Fluoroquinolone, Ciprofloxacin, Ciprofloxacin',\n",
       " 'Streptomycin',\n",
       " 'Apramycin, Gentamicin, Tobramycin, Dibekacin, Netilmicin, Sisomicin, Gentamicin, Tobramycin',\n",
       " 'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Piperacillin, Piperacillin+Tazobactam',\n",
       " 'Amoxicillin, Ampicillin, Aztreonam, Piperacillin, Piperacillin+Tazobactam',\n",
       " 'Amoxicillin, Ampicillin, Aztreonam, Cefepime, Cefotaxime, Ceftazidime, Ceftriaxone, Piperacillin, Ticarcillin',\n",
       " 'Amoxicillin, Ampicillin, Aztreonam, Cefepime, Cefotaxime, Ceftazidime, Ceftriaxone, Piperacillin, Ticarcillin',\n",
       " 'Unknown Beta-lactam',\n",
       " 'Amoxicillin, Ampicillin, Cephalothin, Piperacillin, Ticarcillin',\n",
       " 'Amoxicillin, Amoxicillin+Clavulanic acid, Ampicillin, Ampicillin+Clavulanic acid, Cefepime, Cefixime, Cefotaxime, Cefoxitin, Ceftazidime, Ertapenem, Imipenem, Meropenem, Piperacillin, Piperacillin+Tazobactam, Temocillin',\n",
       " 'Amoxicillin, Ampicillin, Aztreonam, Cefepime, Cefotaxime, Ceftazidime, Ceftriaxone, Piperacillin, Ticarcillin',\n",
       " 'Fosfomycin',\n",
       " 'Fosfomycin',\n",
       " 'Chloramphenicol',\n",
       " 'Ciprofloxacin',\n",
       " 'Ciprofloxacin',\n",
       " 'Chloramphenicol, Nalidixic acid, Ciprofloxacin, Trimethoprim',\n",
       " 'Chloramphenicol, Nalidixic acid, Ciprofloxacin, Trimethoprim',\n",
       " 'Rifampicin',\n",
       " 'Sulfamethoxazole',\n",
       " 'Sulfamethoxazole',\n",
       " 'Sulfamethoxazole',\n",
       " 'Doxycycline, Tetracycline',\n",
       " 'Trimethoprim',\n",
       " 'Trimethoprim']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples=[\"VA1046_2020\",\"VA1184_2021\",\"VA1788_2021\",\n",
    "         \"VA2464_2020\",\"VA418_2022\",\"VA692_2022\",\n",
    "         \"VA1101_2021\",\"VA1565_2021\",\"VA2067_2020\",\n",
    "         \"VA2588_2020\",\"VA585_2022\",\"VA61_2022\"]\n",
    "\n",
    "results = \"/home/npilquinao/BEM-ISP/ResFinder/\"\n",
    "\n",
    "# Predicted resistance genes\n",
    "\n",
    "AMR = pl.read_csv(results + samples[0] + \"/ResFinder_results_tab.txt\",\n",
    "                  separator=\"\\t\",\n",
    "                  has_header=True)\n",
    "AMR = AMR.with_columns(pl.lit(samples[0]).alias(\"Sample\"))\n",
    "                       \n",
    "print(f\"Diferent resistance genes: {AMR.shape[0]}\")\n",
    "\n",
    "resistance = AMR[\"Resistance gene\"].to_list()\n",
    "phenotype = AMR[\"Phenotype\"].to_list()\n",
    "\n",
    "display(AMR)\n",
    "display(resistance, phenotype)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique AMR genes (total): 66\n",
      "\n",
      "AMR genes in VA1046_2020: 24\n",
      "AMR genes in VA1184_2021: 13\n",
      "AMR genes in VA1788_2021: 19\n",
      "AMR genes in VA2464_2020: 14\n",
      "AMR genes in VA418_2022: 20\n",
      "AMR genes in VA692_2022: 13\n",
      "AMR genes in VA1101_2021: 11\n",
      "AMR genes in VA1565_2021: 20\n",
      "AMR genes in VA2067_2020: 12\n",
      "AMR genes in VA2588_2020: 13\n",
      "AMR genes in VA585_2022: 11\n",
      "AMR genes in VA61_2022: 20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (209, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Resistance gene</th><th>Identity</th><th>Alignment Length/Gene Length</th><th>Coverage</th><th>Position in reference</th><th>Contig</th><th>Position in contig</th><th>Phenotype</th><th>Accession no.</th><th>Sample</th></tr><tr><td>str</td><td>f64</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;aph(6)-Id&quot;</td><td>100.0</td><td>&quot;837/837&quot;</td><td>100.0</td><td>&quot;1..838&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Streptomycin&quot;</td><td>&quot;M28829&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;ant(3&#x27;&#x27;)-Ia&quot;</td><td>98.15</td><td>&quot;989/972&quot;</td><td>101.03</td><td>&quot;1..990&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Streptomycin&quot;</td><td>&quot;X02340&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;aac(6&#x27;)-Ib-cr&quot;</td><td>100.0</td><td>&quot;600/600&quot;</td><td>100.0</td><td>&quot;1..601&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Tobramycin, Di‚Ä¶</td><td>&quot;DQ303918&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;aph(3&#x27;&#x27;)-Ib&quot;</td><td>99.0</td><td>&quot;804/804&quot;</td><td>99.88</td><td>&quot;1..805&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Streptomycin&quot;</td><td>&quot;AF321551&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;aac(3)-IIa&quot;</td><td>100.0</td><td>&quot;861/861&quot;</td><td>100.0</td><td>&quot;1..862&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Apramycin, Gen‚Ä¶</td><td>&quot;CP023555&quot;</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;sul1&quot;</td><td>100.0</td><td>&quot;840/840&quot;</td><td>100.0</td><td>&quot;1..841&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Sulfamethoxazo‚Ä¶</td><td>&quot;U12338&quot;</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;dfrA27&quot;</td><td>100.0</td><td>&quot;474/474&quot;</td><td>100.0</td><td>&quot;1..475&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Trimethoprim&quot;</td><td>&quot;FJ459817&quot;</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;dfrA1&quot;</td><td>86.29</td><td>&quot;474/474&quot;</td><td>86.29</td><td>&quot;1..475&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Trimethoprim&quot;</td><td>&quot;AJ238350&quot;</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;dfrA1&quot;</td><td>97.05</td><td>&quot;474/474&quot;</td><td>97.05</td><td>&quot;1..475&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Trimethoprim&quot;</td><td>&quot;AF203818&quot;</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;dfrA16&quot;</td><td>100.0</td><td>&quot;474/474&quot;</td><td>100.0</td><td>&quot;1..475&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA..NA&quot;</td><td>&quot;Trimethoprim&quot;</td><td>&quot;AF174129&quot;</td><td>&quot;VA61_2022&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (209, 10)\n",
       "‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê\n",
       "‚îÇ Resistance ‚îÜ Identity ‚îÜ Alignment ‚îÜ Coverage ‚îÜ ‚Ä¶ ‚îÜ Position  ‚îÜ Phenotype ‚îÜ Accession ‚îÜ Sample    ‚îÇ\n",
       "‚îÇ gene       ‚îÜ ---      ‚îÜ Length/Ge ‚îÜ ---      ‚îÜ   ‚îÜ in contig ‚îÜ ---       ‚îÜ no.       ‚îÜ ---       ‚îÇ\n",
       "‚îÇ ---        ‚îÜ f64      ‚îÜ ne Length ‚îÜ f64      ‚îÜ   ‚îÜ ---       ‚îÜ str       ‚îÜ ---       ‚îÜ str       ‚îÇ\n",
       "‚îÇ str        ‚îÜ          ‚îÜ ---       ‚îÜ          ‚îÜ   ‚îÜ str       ‚îÜ           ‚îÜ str       ‚îÜ           ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ str       ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ           ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°\n",
       "‚îÇ aph(6)-Id  ‚îÜ 100.0    ‚îÜ 837/837   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Streptomy ‚îÜ M28829    ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ ant(3'')-I ‚îÜ 98.15    ‚îÜ 989/972   ‚îÜ 101.03   ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Streptomy ‚îÜ X02340    ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ a          ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ aac(6')-Ib ‚îÜ 100.0    ‚îÜ 600/600   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Tobramyci ‚îÜ DQ303918  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ -cr        ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ n, Dibeka ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin, Amik ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ acin,‚Ä¶    ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ aph(3'')-I ‚îÜ 99.0     ‚îÜ 804/804   ‚îÜ 99.88    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Streptomy ‚îÜ AF321551  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ b          ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin       ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ aac(3)-IIa ‚îÜ 100.0    ‚îÜ 861/861   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Apramycin ‚îÜ CP023555  ‚îÜ VA1046_20 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ , Gentami ‚îÜ           ‚îÜ 20        ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ cin, Tobr ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ amyci‚Ä¶    ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ ‚Ä¶          ‚îÜ ‚Ä¶        ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶        ‚îÜ ‚Ä¶ ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶         ‚îÜ ‚Ä¶         ‚îÇ\n",
       "‚îÇ sul1       ‚îÜ 100.0    ‚îÜ 840/840   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Sulfameth ‚îÜ U12338    ‚îÜ VA61_2022 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ oxazole   ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ dfrA27     ‚îÜ 100.0    ‚îÜ 474/474   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Trimethop ‚îÜ FJ459817  ‚îÜ VA61_2022 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ rim       ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ dfrA1      ‚îÜ 86.29    ‚îÜ 474/474   ‚îÜ 86.29    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Trimethop ‚îÜ AJ238350  ‚îÜ VA61_2022 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ rim       ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ dfrA1      ‚îÜ 97.05    ‚îÜ 474/474   ‚îÜ 97.05    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Trimethop ‚îÜ AF203818  ‚îÜ VA61_2022 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ rim       ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îÇ dfrA16     ‚îÜ 100.0    ‚îÜ 474/474   ‚îÜ 100.0    ‚îÜ ‚Ä¶ ‚îÜ NA..NA    ‚îÜ Trimethop ‚îÜ AF174129  ‚îÜ VA61_2022 ‚îÇ\n",
       "‚îÇ            ‚îÜ          ‚îÜ           ‚îÜ          ‚îÜ   ‚îÜ           ‚îÜ rim       ‚îÜ           ‚îÜ           ‚îÇ\n",
       "‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['qnrB6',\n",
       " 'blaCTX-M-3',\n",
       " 'blaVIM-1',\n",
       " \"aac(6')-Il\",\n",
       " 'blaACT-6',\n",
       " 'blaACT-5',\n",
       " 'cmlA1',\n",
       " 'fosA5',\n",
       " 'blaSHV-81',\n",
       " 'sul1',\n",
       " 'blaACT-14',\n",
       " 'blaSHV-40',\n",
       " 'qnrB19',\n",
       " 'blaKPC-2',\n",
       " 'dfrA16',\n",
       " \"aac(6')-Ib3\",\n",
       " 'aac(3)-IId',\n",
       " 'blaCTX-M-2',\n",
       " 'mph(E)',\n",
       " 'blaOXA-10',\n",
       " 'aadA2',\n",
       " \"aac(6')-Ib\",\n",
       " \"ant(3'')-Ia\",\n",
       " 'blaSHV-110',\n",
       " 'ARR-3',\n",
       " 'armA',\n",
       " 'msr(E)',\n",
       " 'aph(6)-Id',\n",
       " 'blaNDM-7',\n",
       " 'blaNDM-1',\n",
       " 'blaOXA-1',\n",
       " \"aph(3'')-Ib\",\n",
       " 'blaVIM-2',\n",
       " 'blaSHV-11',\n",
       " 'blaACT-7',\n",
       " 'blaACT-10',\n",
       " 'dfrA12',\n",
       " 'blaSHV-2a',\n",
       " 'blaCTX-M-15',\n",
       " \"ant(2'')-Ia\",\n",
       " 'dfrA14',\n",
       " 'aac(3)-IIa',\n",
       " 'dfrA27',\n",
       " 'dfrA1',\n",
       " 'fosA',\n",
       " \"aac(6')-Ib-cr\",\n",
       " 'qnrA1',\n",
       " 'blaSHV-27',\n",
       " 'blaSHV-190',\n",
       " 'blaCTX-M-9',\n",
       " 'fosA6',\n",
       " 'tet(D)',\n",
       " 'aadA16',\n",
       " 'blaSHV-148',\n",
       " 'blaACT-16',\n",
       " 'OqxB',\n",
       " 'blaACT-15',\n",
       " 'tet(A)',\n",
       " 'OqxA',\n",
       " 'aadA1',\n",
       " 'blaSHV-1',\n",
       " 'blaSHV-158',\n",
       " 'qnrB1',\n",
       " 'blaKPC-3',\n",
       " 'blaTEM-1B',\n",
       " 'sul2']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples=[\"VA1046_2020\",\"VA1184_2021\",\"VA1788_2021\",\n",
    "         \"VA2464_2020\",\"VA418_2022\",\"VA692_2022\",\n",
    "         \"VA1101_2021\",\"VA1565_2021\",\"VA2067_2020\",\n",
    "         \"VA2588_2020\",\"VA585_2022\",\"VA61_2022\"]\n",
    "\n",
    "results = \"/home/npilquinao/BEM-ISP/ResFinder/\"\n",
    "\n",
    "AMR = pl.DataFrame()\n",
    "\n",
    "for sample in samples:\n",
    "    sample_df = pl.read_csv(results + sample + \"/ResFinder_results_tab.txt\",\n",
    "                            separator=\"\\t\",\n",
    "                            has_header=True)\n",
    "    sample_df = sample_df.with_columns(pl.lit(sample).alias(\"Sample\"))\n",
    "    AMR = AMR.vstack(sample_df)\n",
    "\n",
    "print(f\"Unique AMR genes (total): {AMR.unique('Resistance gene').shape[0]}\")\n",
    "print(\"\")\n",
    "\n",
    "for sample in samples:\n",
    "    sample_df = AMR.filter(pl.col(\"Sample\") == sample)\n",
    "    print(f\"AMR genes in {sample}: {sample_df.unique('Resistance gene').shape[0]}\")\n",
    "\n",
    "display(AMR)\n",
    "\n",
    "display((AMR[\"Resistance gene\"].unique()).to_list())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Filter for carbapenemases genes only__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensions: (40, 4)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['VA1046_2020',\n",
       " 'VA1046_2020',\n",
       " 'VA1046_2020',\n",
       " 'VA1046_2020',\n",
       " 'VA1184_2021',\n",
       " 'VA1184_2021',\n",
       " 'VA1184_2021',\n",
       " 'VA1788_2021',\n",
       " 'VA1788_2021',\n",
       " 'VA2464_2020',\n",
       " 'VA2464_2020',\n",
       " 'VA2464_2020',\n",
       " 'VA418_2022',\n",
       " 'VA418_2022',\n",
       " 'VA692_2022',\n",
       " 'VA692_2022',\n",
       " 'VA692_2022',\n",
       " 'VA1101_2021',\n",
       " 'VA1101_2021',\n",
       " 'VA1101_2021',\n",
       " 'VA1101_2021',\n",
       " 'VA1565_2021',\n",
       " 'VA1565_2021',\n",
       " 'VA1565_2021',\n",
       " 'VA1565_2021',\n",
       " 'VA1565_2021',\n",
       " 'VA2067_2020',\n",
       " 'VA2067_2020',\n",
       " 'VA2067_2020',\n",
       " 'VA2588_2020',\n",
       " 'VA2588_2020',\n",
       " 'VA2588_2020',\n",
       " 'VA2588_2020',\n",
       " 'VA585_2022',\n",
       " 'VA585_2022',\n",
       " 'VA585_2022',\n",
       " 'VA61_2022',\n",
       " 'VA61_2022',\n",
       " 'VA61_2022',\n",
       " 'VA61_2022']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (40, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Resistance gene</th><th>Identity</th><th>Phenotype</th><th>Sample</th></tr><tr><td>str</td><td>f64</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;blaSHV-2a&quot;</td><td>91.52</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;blaSHV-27&quot;</td><td>85.71</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;blaSHV-110&quot;</td><td>99.54</td><td>&quot;Unknown Beta-l‚Ä¶</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;blaNDM-7&quot;</td><td>100.0</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA1046_2020&quot;</td></tr><tr><td>&quot;blaACT-5&quot;</td><td>98.87</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA1184_2021&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;blaSHV-190&quot;</td><td>99.88</td><td>&quot;Unknown Beta-l‚Ä¶</td><td>&quot;VA585_2022&quot;</td></tr><tr><td>&quot;blaACT-7&quot;</td><td>99.48</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;blaACT-15&quot;</td><td>98.08</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;blaKPC-2&quot;</td><td>100.0</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA61_2022&quot;</td></tr><tr><td>&quot;blaVIM-1&quot;</td><td>100.0</td><td>&quot;Amoxicillin, A‚Ä¶</td><td>&quot;VA61_2022&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (40, 4)\n",
       "‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê\n",
       "‚îÇ Resistance gene ‚îÜ Identity ‚îÜ Phenotype                         ‚îÜ Sample      ‚îÇ\n",
       "‚îÇ ---             ‚îÜ ---      ‚îÜ ---                               ‚îÜ ---         ‚îÇ\n",
       "‚îÇ str             ‚îÜ f64      ‚îÜ str                               ‚îÜ str         ‚îÇ\n",
       "‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°\n",
       "‚îÇ blaSHV-2a       ‚îÜ 91.52    ‚îÜ Amoxicillin, Ampicillin, Aztreon‚Ä¶ ‚îÜ VA1046_2020 ‚îÇ\n",
       "‚îÇ blaSHV-27       ‚îÜ 85.71    ‚îÜ Amoxicillin, Ampicillin, Aztreon‚Ä¶ ‚îÜ VA1046_2020 ‚îÇ\n",
       "‚îÇ blaSHV-110      ‚îÜ 99.54    ‚îÜ Unknown Beta-lactam               ‚îÜ VA1046_2020 ‚îÇ\n",
       "‚îÇ blaNDM-7        ‚îÜ 100.0    ‚îÜ Amoxicillin, Amoxicillin+Clavula‚Ä¶ ‚îÜ VA1046_2020 ‚îÇ\n",
       "‚îÇ blaACT-5        ‚îÜ 98.87    ‚îÜ Amoxicillin, Amoxicillin+Clavula‚Ä¶ ‚îÜ VA1184_2021 ‚îÇ\n",
       "‚îÇ ‚Ä¶               ‚îÜ ‚Ä¶        ‚îÜ ‚Ä¶                                 ‚îÜ ‚Ä¶           ‚îÇ\n",
       "‚îÇ blaSHV-190      ‚îÜ 99.88    ‚îÜ Unknown Beta-lactam               ‚îÜ VA585_2022  ‚îÇ\n",
       "‚îÇ blaACT-7        ‚îÜ 99.48    ‚îÜ Amoxicillin, Amoxicillin+Clavula‚Ä¶ ‚îÜ VA61_2022   ‚îÇ\n",
       "‚îÇ blaACT-15       ‚îÜ 98.08    ‚îÜ Amoxicillin, Amoxicillin+Clavula‚Ä¶ ‚îÜ VA61_2022   ‚îÇ\n",
       "‚îÇ blaKPC-2        ‚îÜ 100.0    ‚îÜ Amoxicillin, Amoxicillin+Clavula‚Ä¶ ‚îÜ VA61_2022   ‚îÇ\n",
       "‚îÇ blaVIM-1        ‚îÜ 100.0    ‚îÜ Amoxicillin, Amoxicillin+Clavula‚Ä¶ ‚îÜ VA61_2022   ‚îÇ\n",
       "‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "carbapenemase = [\"ACT\",\"KPC\",\"SHV\",\"NDM\",\"VIM\"]\n",
    "\n",
    "AMR_filtered = AMR.filter(pl.col('Resistance gene').map_elements(lambda value: any(word in value for word in carbapenemase), return_dtype=bool))\n",
    "AMR_filtered = AMR_filtered.drop([\"Alignment Length/Gene Length\",\n",
    "                                 \"Coverage\",\n",
    "                                 \"Position in reference\",\n",
    "                                 \"Contig\",\n",
    "                                 \"Position in contig\",\n",
    "                                 \"Accession no.\"])\n",
    "print(f\"Dimensions: {AMR_filtered.shape}\")\n",
    "display(AMR_filtered[\"Sample\"].to_list())\n",
    "display(AMR_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Seg√∫n chatGPT para:\n",
    "\n",
    "blaVIM-1\n",
    "blaACT-6\n",
    "blaACT-5\n",
    "blaACT-14\n",
    "blaKPC-2\n",
    "blaSHV-110\n",
    "blaNDM-7\n",
    "blaNDM-1\n",
    "blaVIM-2\n",
    "blaACT-7\n",
    "blaACT-10\n",
    "blaSHV-2a\n",
    "blaKPC-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "pl.DataFrame.write_csv(AMR_filtered, \"~/BEM-ISP/ResFinder/only_carbapenemase.tsv\", separator=\"\\t\")\n",
    "pl.DataFrame.write_csv(AMR, \"~/BEM-ISP/ResFinder/AMR_ResFinder.tsv\", separator=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Desorden__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phenotype results\n",
    "\n",
    "phenotype = pl.read_csv(\"/home/npilquinao/BEM-ISP/ResFinder/VA61_2022/pheno_table.txt\",\n",
    "                        separator=\"\\t\",\n",
    "                        skip_rows=16,\n",
    "                        has_header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/npilquinao/miniconda3/envs/ISP/lib/python3.11/site-packages/resfinder/__init__.py\n"
     ]
    }
   ],
   "source": [
    "print(resfinder.__file__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "amr_abbreviations.md\t  __init__.py  run_resfinder.py\n",
      "cge\t\t\t  __main__.py  species_abbreviations.md\n",
      "environment_variables.md  __pycache__\n",
      "\n",
      "#!/usr/bin/env python3\n",
      "import io\n",
      "import sys\n",
      "import os\n",
      "import subprocess\n",
      "from argparse import ArgumentParser\n",
      "import pickle\n",
      "import json\n",
      "import hashlib\n",
      "\n",
      "from cgelib.output.result import Result\n",
      "from cgelib.utils.loaders_mixin import LoadersMixin\n",
      "from cgelib.utils.pliers_mixin import PliersMixin\n",
      "\n",
      "from resfinder.cge.config import Config\n",
      "from resfinder.cge.resfinder import ResFinder\n",
      "from resfinder.cge.pointfinder import PointFinder\n",
      "from resfinder.cge.output.std_results import ResFinderResultHandler\n",
      "from resfinder.cge.output.std_results import PointFinderResultHandler\n",
      "\n",
      "#  Modules used to create the extended ResFinder output (phenotype output)\n",
      "from resfinder.cge.phenotype2genotype.isolate import Isolate\n",
      "from resfinder.cge.phenotype2genotype.res_profile import PhenoDB\n",
      "from resfinder.cge.phenotype2genotype.res_sumtable import ResSumTable\n",
      "from resfinder.cge.phenotype2genotype.res_sumtable import PanelNameError\n",
      "\n",
      "from resfinder import __version__\n",
      "\n",
      "\n",
      "def eprint(*args, **kwargs):\n",
      "    print(*args, file=sys.stderr, **kwargs)\n",
      "\n",
      "\n",
      "def get_software_exec_res(conf: Config) -> dict:\n",
      "    software_exec_res = {\n",
      "        \"type\": \"software_exec\",\n",
      "        \"command\": \" \".join(sys.argv),\n",
      "        \"parameters\": get_call_parameters(conf)\n",
      "    }\n",
      "    software_exec_res[\"key\"] = hashlib.sha1(\n",
      "        bytes(software_exec_res[\"command\"], 'UTF-8')).hexdigest()\n",
      "    return software_exec_res\n",
      "\n",
      "\n",
      "def get_call_parameters(conf: Config) -> dict:\n",
      "    parameters = vars(conf).copy()\n",
      "    del (parameters['amr_abbreviations'])\n",
      "    return parameters\n",
      "\n",
      "\n",
      "def main():\n",
      "\n",
      "    # version = read_version(\"VERSION\")\n",
      "    version = __version__\n",
      "\n",
      "    ##########################################################################\n",
      "    # PARSE COMMAND LINE OPTIONS\n",
      "    ##########################################################################\n",
      "\n",
      "    parser = ArgumentParser(allow_abbrev=False)\n",
      "\n",
      "    # General options\n",
      "    parser.add_argument(\"-ifa\", \"--inputfasta\",\n",
      "                        help=\"Input fasta file.\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-ifq\", \"--inputfastq\",\n",
      "                        help=(\"Input fastq file(s). Assumed to be single-end \"\n",
      "                              \"fastq if only one file is provided, and assumed\"\n",
      "                              \" to be paired-end data if two files are \"\n",
      "                              \"provided.\"),\n",
      "                        nargs=\"+\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--nanopore\",\n",
      "                        action=\"store_true\",\n",
      "                        dest=\"nanopore\",\n",
      "                        help=\"If nanopore data is used\",\n",
      "                        default=False)\n",
      "    parser.add_argument(\"-o\", \"--outputPath\",\n",
      "                        help=(\"Output directory. If it doesn't exist, it will \"\n",
      "                              \"be created.\"),\n",
      "                        required=True,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-j\", \"--out_json\",\n",
      "                        help=(\"Specify JSON filename and output directory. If \"\n",
      "                              \"the directory doesn't exist, it will be \"\n",
      "                              \"created.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-b\", \"--blastPath\",\n",
      "                        help=\"Path to blastn\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-k\", \"--kmaPath\",\n",
      "                        help=\"Path to KMA\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-s\", \"--species\",\n",
      "                        help=\"Species in the sample\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--ignore_missing_species\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"If set, species is provided and --point flag \"\n",
      "                              \"is set, will not throw an error if no database \"\n",
      "                              \"is found for the provided species. If species \"\n",
      "                              \"is not found. Point mutations will silently \"\n",
      "                              \"be ignored.\"),\n",
      "                        default=False)\n",
      "    parser.add_argument(\"--output_aln\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"will add the alignments in the json output.\",\n",
      "                        default=False)\n",
      "\n",
      "    # Acquired resistance options\n",
      "    parser.add_argument(\"-db_res\", \"--db_path_res\",\n",
      "                        help=(\"Path to the databases for ResFinder.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_res_kma\", \"--db_path_res_kma\",\n",
      "                        help=(\"Path to the ResFinder databases indexed with \"\n",
      "                              \"KMA. Defaults to the value of the --db_res \"\n",
      "                              \"flag.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-acq\", \"--acquired\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Run resfinder for acquired resistance genes\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-ao\", \"--acq_overlap\",\n",
      "                        help=\"Genes are allowed to overlap this number of\\\n",
      "                              nucleotides. Default: {}.\".format(\n",
      "                            Config.DEFAULT_VALS[\"acq_overlap\"]),\n",
      "                        type=int,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-l\", \"--min_cov\",\n",
      "                        help=(\"Minimum (breadth-of) coverage of ResFinder \"\n",
      "                              \"within the range 0-1.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-t\", \"--threshold\",\n",
      "                        help=(\"Threshold for identity of ResFinder within the \"\n",
      "                              \"range 0-1.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    # Disinfectant resistance options\n",
      "    parser.add_argument(\"-d\", \"--disinfectant\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Run resfinder for disinfectant resistance genes\",\n",
      "                        default=False)\n",
      "    parser.add_argument(\"-db_disinf\", \"--db_path_disinf\",\n",
      "                        help=(\"Path to the databases for DisinFinder.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_disinf_kma\", \"--db_path_disinf_kma\",\n",
      "                        help=(\"Path to the DisinFinder databases indexed with \"\n",
      "                              \"KMA. Defaults to the value of the --db_res \"\n",
      "                              \"flag.\"),\n",
      "                        default=None)\n",
      "\n",
      "    # Point resistance option\n",
      "    parser.add_argument(\"-c\", \"--point\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Run pointfinder for chromosomal mutations\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_point\", \"--db_path_point\",\n",
      "                        help=\"Path to the databases for PointFinder\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_point_kma\", \"--db_path_point_kma\",\n",
      "                        help=(\"Path to the PointFinder databases indexed with \"\n",
      "                              \"KMA. Defaults to the value of the \"\n",
      "                              \"--db_path_point flag.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-g\", \"--specific_gene\",\n",
      "                        nargs='+',\n",
      "                        help=\"Specify genes existing in the database to \\\n",
      "                              search for - if none is specified all genes are \\\n",
      "                              included in the search.\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-u\", \"--unknown_mut\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"Show all mutations found even if in unknown to \"\n",
      "                              \"the resistance database\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-l_p\", \"--min_cov_point\",\n",
      "                        help=(\"Minimum (breadth-of) coverage of Pointfinder \"\n",
      "                              \"within the range 0-1. If None is selected, the \"\n",
      "                              \"minimum coverage of ResFinder will be used.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-t_p\", \"--threshold_point\",\n",
      "                        help=(\"Threshold for identity of Pointfinder within \"\n",
      "                              \"the range 0-1. If None is selected, the minimum\"\n",
      "                              \" coverage of ResFinder will be used.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--ignore_indels\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"Ignore frameshift-causing indels in \"\n",
      "                              \"Pointfinder.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--ignore_stop_codons\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Ignore premature stop codons in Pointfinder.\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-v\", \"--version\", action=\"version\",\n",
      "                        version=__version__,\n",
      "                        help=\"Show program's version number and exit\")\n",
      "\n",
      "    # Temporary option only available temporary\n",
      "    parser.add_argument(\"--pickle\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"Create a pickle dump of the Isolate object. \"\n",
      "                              \"Currently needed in the CGE webserver. \"\n",
      "                              \"Dependency and this option is being removed.\"),\n",
      "                        default=False)\n",
      "\n",
      "    args = parser.parse_args()\n",
      "\n",
      "    # Parse and check all arguments and expected files.\n",
      "    conf = Config(args)\n",
      "\n",
      "    # Initialise result dict\n",
      "    std_result = Result.init_software_result(\n",
      "        name=\"ResFinder\",\n",
      "        gitdir=f\"{conf.resfinder_root}/../../\")\n",
      "\n",
      "    init_result_data = {\n",
      "        \"provided_species\": conf.species,\n",
      "        \"software_version\": __version__,\n",
      "        \"key\": f\"ResFinder-{__version__}\",\n",
      "    }\n",
      "    std_result.add(**init_result_data)\n",
      "\n",
      "    if conf.acquired:\n",
      "        std_result.init_database(\"ResFinder\", conf.db_path_res)\n",
      "    if conf.point:\n",
      "        std_result.init_database(\"PointFinder\", conf.db_path_point_root)\n",
      "    if conf.disinf:\n",
      "        std_result.init_database(\"DisinFinder\", conf.db_path_disinf)\n",
      "\n",
      "    std_result.add_class(cl=\"software_executions\", **\n",
      "                         get_software_exec_res(conf))\n",
      "\n",
      "    # Load genotype to phenotype database\n",
      "    res_pheno_db = PhenoDB(\n",
      "        abclassdef_file=conf.abclassdef_file,\n",
      "        acquired_file=conf.phenotype_file,\n",
      "        point_file=conf.point_file,\n",
      "        disinf_file=conf.disinf_file,\n",
      "        disclassdef_file=conf.disclassdef_file\n",
      "    )\n",
      "\n",
      "    ##########################################################################\n",
      "    # ResFinder\n",
      "    ##########################################################################\n",
      "\n",
      "    if conf.acquired is True:\n",
      "\n",
      "        blast_results = None\n",
      "        kma_run = None\n",
      "\n",
      "        # Actually running ResFinder (for acquired resistance)\n",
      "        acquired_finder = ResFinder(db_conf_file=conf.db_config_file,\n",
      "                                    db_path=conf.db_path_res,\n",
      "                                    pheno_file=conf.phenotype_file,\n",
      "                                    notes=conf.db_notes_file,\n",
      "                                    db_path_kma=conf.db_path_res_kma)\n",
      "\n",
      "        if (conf.inputfasta):\n",
      "            blast_results = acquired_finder.blast(\n",
      "                inputfile=conf.inputfasta,\n",
      "                out_path=conf.outPath_res_blast,\n",
      "                min_cov=conf.rf_gene_cov,\n",
      "                threshold=conf.rf_gene_id,\n",
      "                blast=conf.blast,\n",
      "                allowed_overlap=conf.rf_overlap\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            acquired_finder.write_results(out_path=conf.outputPath,\n",
      "                                          result=blast_results,\n",
      "                                          res_type=ResFinder.TYPE_BLAST,\n",
      "                                          software=\"ResFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       blast_results,\n",
      "                                                       \"ResFinder\",\n",
      "                                                       conf)\n",
      "\n",
      "        else:\n",
      "            kma_nano_args = None\n",
      "            if conf.nanopore:\n",
      "                kma_nano_args = '-ont -md 5'\n",
      "\n",
      "            kma_run = acquired_finder.kma(\n",
      "                inputfile_1=conf.inputfastq_1,\n",
      "                inputfile_2=conf.inputfastq_2,\n",
      "                out_path=conf.outPath_res_kma,\n",
      "                db_path_kma=conf.db_path_res_kma,\n",
      "                min_cov=conf.rf_gene_cov,\n",
      "                threshold=conf.rf_gene_id,\n",
      "                kma_path=conf.kma,\n",
      "                databases=acquired_finder.databases,\n",
      "                sample_name=\"\",\n",
      "                kma_cge=True,\n",
      "                kma_apm=\"p\",\n",
      "                kma_1t1=True,\n",
      "                kma_add_args=kma_nano_args\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            acquired_finder.write_results(out_path=conf.outputPath,\n",
      "                                          result=kma_run.results,\n",
      "                                          res_type=ResFinder.TYPE_KMA,\n",
      "                                          software=\"ResFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       kma_run,\n",
      "                                                       \"ResFinder\",\n",
      "                                                       conf)\n",
      "    ##########################################################################\n",
      "    # DisinFinder\n",
      "    ##########################################################################\n",
      "    if (conf.disinf is True):\n",
      "\n",
      "        blast_results = None\n",
      "        kma_run = None\n",
      "\n",
      "        # Actually running DisinFinder (for disinfectant resistance)\n",
      "        disinf_finder = ResFinder(db_conf_file=conf.db_config_disinf_file,\n",
      "                                  db_path=conf.db_path_disinf,\n",
      "                                  pheno_file=conf.disinf_file,\n",
      "                                  notes=conf.db_notes_disinf_file,\n",
      "                                  db_path_kma=conf.db_path_disinf_kma)\n",
      "\n",
      "        if (conf.inputfasta):\n",
      "            blast_results = disinf_finder.blast(\n",
      "                inputfile=conf.inputfasta,\n",
      "                out_path=conf.outPath_disinf_blast,\n",
      "                min_cov=conf.dis_gene_cov,\n",
      "                threshold=conf.dis_gene_id,\n",
      "                blast=conf.blast,\n",
      "                allowed_overlap=conf.dis_overlap\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            disinf_finder.write_results(out_path=conf.outputPath,\n",
      "                                        result=blast_results,\n",
      "                                        res_type=ResFinder.TYPE_BLAST,\n",
      "                                        software=\"DisinFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       blast_results,\n",
      "                                                       \"DisinFinder\",\n",
      "                                                       conf)\n",
      "\n",
      "        else:\n",
      "            kma_nano_args = None\n",
      "            if conf.nanopore:\n",
      "                kma_nano_args = '-ont -md 5'\n",
      "\n",
      "            kma_run = disinf_finder.kma(\n",
      "                inputfile_1=conf.inputfastq_1,\n",
      "                inputfile_2=conf.inputfastq_2,\n",
      "                out_path=conf.outPath_disinf_kma,\n",
      "                db_path_kma=conf.db_path_disinf_kma,\n",
      "                min_cov=conf.dis_gene_cov,\n",
      "                threshold=conf.dis_gene_id,\n",
      "                kma_path=conf.kma,\n",
      "                databases=disinf_finder.databases,\n",
      "                sample_name=\"\",\n",
      "                kma_cge=True,\n",
      "                kma_apm=\"p\",\n",
      "                kma_1t1=True,\n",
      "                kma_add_args=kma_nano_args\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            disinf_finder.write_results(out_path=conf.outputPath,\n",
      "                                        result=kma_run.results,\n",
      "                                        res_type=ResFinder.TYPE_KMA,\n",
      "                                        software=\"DisinFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       kma_run,\n",
      "                                                       \"DisinFinder\",\n",
      "                                                       conf)\n",
      "    ##########################################################################\n",
      "    # PointFinder\n",
      "    ##########################################################################\n",
      "\n",
      "    if (conf.point):\n",
      "\n",
      "        blast_results = None\n",
      "        kma_run = None\n",
      "\n",
      "        finder = PointFinder(db_path=conf.db_path_point,\n",
      "                             species=conf.species_dir,\n",
      "                             gene_list=conf.specific_gene,\n",
      "                             ignore_indels=conf.ignore_indels,\n",
      "                             ignore_stop_codons=conf.ignore_stop_codons)\n",
      "\n",
      "        if (conf.inputfasta):\n",
      "\n",
      "            method = PointFinder.TYPE_BLAST\n",
      "\n",
      "            blast_run = finder.blast(inputfile=conf.inputfasta,\n",
      "                                     out_path=conf.outPath_point_blast,\n",
      "                                     min_cov=0.01,  # Sorts on coverage later\n",
      "                                     threshold=conf.pf_gene_id,\n",
      "                                     blast=conf.blast,\n",
      "                                     cut_off=False)\n",
      "            results = blast_run.results\n",
      "\n",
      "        else:\n",
      "            kma_nano_args = None\n",
      "            if conf.nanopore:\n",
      "                kma_nano_args = '-ont -md 5'\n",
      "            \n",
      "            method = PointFinder.TYPE_KMA\n",
      "            \n",
      "            kma_run = finder.kma(inputfile_1=conf.inputfastq_1,\n",
      "                                    inputfile_2=conf.inputfastq_2,\n",
      "                                    out_path=conf.outPath_point_kma,\n",
      "                                    db_path_kma=conf.db_path_point,\n",
      "                                    databases=[conf.species_dir],\n",
      "                                    min_cov=0.01,  # Sorts on coverage later\n",
      "                                    threshold=conf.pf_gene_id,\n",
      "                                    kma_path=conf.kma,\n",
      "                                    sample_name=conf.sample_name,\n",
      "                                    kma_cge=True,\n",
      "                                    kma_apm=\"p\",\n",
      "                                    kma_1t1=True,\n",
      "                                    kma_add_args=kma_nano_args)\n",
      "\n",
      "            results = kma_run.results\n",
      "        \n",
      "        PointFinderResultHandler.rename_promoter_results(results, method)\n",
      "        \n",
      "        if (conf.specific_gene):\n",
      "            results = PointFinder.discard_from_dict(\n",
      "                in_dict=results, wanted_list=conf.specific_gene)\n",
      "\n",
      "        if (method == PointFinder.TYPE_BLAST):\n",
      "            results_pnt = finder.find_best_seqs(results, conf.pf_gene_cov)\n",
      "        else:\n",
      "            results_pnt = results[finder.species]\n",
      "            if (results_pnt == \"No hit found\"):\n",
      "                results_pnt = {}\n",
      "            else:\n",
      "                results_pnt[\"excluded\"] = results[\"excluded\"]\n",
      "\n",
      "        # DEPRECATED\n",
      "        # mutations in raw reads is only found with write_results.\n",
      "        # TODO: make a write method that depends on the json output\n",
      "        finder.write_results(\n",
      "            out_path=conf.outputPath, result=results, res_type=method,\n",
      "            unknown_flag=conf.unknown_mut, min_cov=conf.pf_gene_cov,\n",
      "            perc_iden=conf.pf_gene_id)\n",
      "\n",
      "        if not conf.unknown_mut:\n",
      "            results_pnt = PointFinder.discard_unknown_muts(\n",
      "                results_pnt=results_pnt, phenodb=res_pheno_db, method=method)\n",
      "\n",
      "        PointFinderResultHandler.standardize_results(std_result,\n",
      "                                                     results_pnt,\n",
      "                                                     \"PointFinder\")\n",
      "\n",
      "    ##########################################################################\n",
      "    # Phenotype to genotype\n",
      "    ##########################################################################\n",
      "\n",
      "    # Isolate object store results\n",
      "    isolate = Isolate(name=conf.sample_name, species=conf.species,\n",
      "                      amr_panel_file=conf.db_panels_file)\n",
      "\n",
      "    if (conf.acquired or conf.disinf):\n",
      "        isolate.load_finder_results(std_table=std_result,\n",
      "                                    phenodb=res_pheno_db,\n",
      "                                    type=\"seq_regions\")\n",
      "    if (conf.point):\n",
      "        isolate.load_finder_results(std_table=std_result,\n",
      "                                    phenodb=res_pheno_db,\n",
      "                                    type=\"seq_variations\")\n",
      "\n",
      "    isolate.calc_res_profile(res_pheno_db)\n",
      "    ResFinderResultHandler.load_res_profile(std_result, isolate,\n",
      "                                            conf.amr_abbreviations)\n",
      "\n",
      "    if (conf.out_json):\n",
      "        std_result_file = conf.out_json\n",
      "    else:\n",
      "        std_result_file = \"{}/{}.json\".format(\n",
      "            conf.outputPath, conf.sample_name.replace(\"_R1\", \"\").split(\".\")[0])\n",
      "    with open(std_result_file, 'w') as fh:\n",
      "        fh.write(std_result.json_dumps())\n",
      "\n",
      "    # Create and write the downloadable tab file\n",
      "    pheno_profile_str = isolate.profile_to_str_table(header=True)\n",
      "\n",
      "    # TODO: REMOVE THE NEED FOR THE PICKLED FILE\n",
      "    if (conf.pickle):\n",
      "        isolate_pickle = open(\"{}/isolate.p\".format(conf.outputPath), \"wb\")\n",
      "        pickle.dump(isolate, isolate_pickle, protocol=2)\n",
      "\n",
      "    pheno_table_file = \"{}/pheno_table.txt\".format(conf.outputPath)\n",
      "    with open(pheno_table_file, 'w') as fh:\n",
      "        fh.write(pheno_profile_str)\n",
      "\n",
      "    if (conf.species is not None):\n",
      "        # Apply AMR panel\n",
      "        input_amr_panels = \"{}/phenotype_panels.txt\".format(conf.db_path_res)\n",
      "        res_sum_table = ResSumTable(pheno_profile_str)\n",
      "        res_sum_table.load_amr_panels(input_amr_panels)\n",
      "\n",
      "        try:\n",
      "            panel_profile_str = res_sum_table.get_amr_panel_str(\n",
      "                panel_name_raw=conf.species, header=True)\n",
      "        # If specified species does not have an associated panel, just ignore it\n",
      "        # and exit.\n",
      "        except PanelNameError:\n",
      "            if conf.species != 'other':\n",
      "                eprint(\"Warning: No panel was detected for the species: {}\"\n",
      "                       .format(conf.species))\n",
      "            sys.exit()\n",
      "\n",
      "        amr_panel_filename = conf.species.replace(\" \", \"_\")\n",
      "\n",
      "        panel_tabel_file = (\"{tablename}_{species}.txt\"\n",
      "                            .format(tablename=pheno_table_file[:-4],\n",
      "                                    species=amr_panel_filename))\n",
      "\n",
      "        with open(panel_tabel_file, \"w\") as fh:\n",
      "            fh.write(panel_profile_str)\n",
      "\n",
      "    return 0\n",
      "\n",
      "\n",
      "if __name__ == '__main__':\n",
      "    sys.exit(main())\n"
     ]
    }
   ],
   "source": [
    "!ls /home/npilquinao/miniconda3/envs/ISP/lib/python3.11/site-packages/resfinder/\n",
    "print(\"\")\n",
    "!cat /home/npilquinao/miniconda3/envs/ISP/lib/python3.11/site-packages/resfinder/run_resfinder.py"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ISP",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
