{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __ResFinder__\n",
    "\n",
    "Search carbapenemase gen occurrences using illumina reads directly\n",
    "\n",
    "__Dependences__\n",
    "\n",
    "- KMA (bioconda)\n",
    "- BLAST (bioconda)\n",
    "- resfinder (pip)\n",
    "\n",
    "Databases\n",
    "\n",
    "(in ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/)\n",
    "\n",
    "- git clone https://bitbucket.org/genomicepidemiology/resfinder_db/\n",
    "- git clone https://bitbucket.org/genomicepidemiology/pointfinder_db/\n",
    "- git clone https://bitbucket.org/genomicepidemiology/disinfinder_db/\n",
    "\n",
    "Por alguna razon, pese a que resfinder es un programa en python, estÃ¡ construido solamente para ejecutarlo mediante CLI xD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: __main__.py [-h] [-ifa INPUTFASTA] [-ifq INPUTFASTQ [INPUTFASTQ ...]]\n",
      "                   [--nanopore] -o OUTPUTPATH [-j OUT_JSON] [-b BLASTPATH]\n",
      "                   [-k KMAPATH] [-s SPECIES] [--ignore_missing_species]\n",
      "                   [--output_aln] [-db_res DB_PATH_RES]\n",
      "                   [-db_res_kma DB_PATH_RES_KMA] [-acq] [-ao ACQ_OVERLAP]\n",
      "                   [-l MIN_COV] [-t THRESHOLD] [-d]\n",
      "                   [-db_disinf DB_PATH_DISINF]\n",
      "                   [-db_disinf_kma DB_PATH_DISINF_KMA] [-c]\n",
      "                   [-db_point DB_PATH_POINT] [-db_point_kma DB_PATH_POINT_KMA]\n",
      "                   [-g SPECIFIC_GENE [SPECIFIC_GENE ...]] [-u]\n",
      "                   [-l_p MIN_COV_POINT] [-t_p THRESHOLD_POINT]\n",
      "                   [--ignore_indels] [--ignore_stop_codons] [-v] [--pickle]\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  -ifa INPUTFASTA, --inputfasta INPUTFASTA\n",
      "                        Input fasta file.\n",
      "  -ifq INPUTFASTQ [INPUTFASTQ ...], --inputfastq INPUTFASTQ [INPUTFASTQ ...]\n",
      "                        Input fastq file(s). Assumed to be single-end fastq if\n",
      "                        only one file is provided, and assumed to be paired-\n",
      "                        end data if two files are provided.\n",
      "  --nanopore            If nanopore data is used\n",
      "  -o OUTPUTPATH, --outputPath OUTPUTPATH\n",
      "                        Output directory. If it doesn't exist, it will be\n",
      "                        created.\n",
      "  -j OUT_JSON, --out_json OUT_JSON\n",
      "                        Specify JSON filename and output directory. If the\n",
      "                        directory doesn't exist, it will be created.\n",
      "  -b BLASTPATH, --blastPath BLASTPATH\n",
      "                        Path to blastn\n",
      "  -k KMAPATH, --kmaPath KMAPATH\n",
      "                        Path to KMA\n",
      "  -s SPECIES, --species SPECIES\n",
      "                        Species in the sample\n",
      "  --ignore_missing_species\n",
      "                        If set, species is provided and --point flag is set,\n",
      "                        will not throw an error if no database is found for\n",
      "                        the provided species. If species is not found. Point\n",
      "                        mutations will silently be ignored.\n",
      "  --output_aln          will add the alignments in the json output.\n",
      "  -db_res DB_PATH_RES, --db_path_res DB_PATH_RES\n",
      "                        Path to the databases for ResFinder.\n",
      "  -db_res_kma DB_PATH_RES_KMA, --db_path_res_kma DB_PATH_RES_KMA\n",
      "                        Path to the ResFinder databases indexed with KMA.\n",
      "                        Defaults to the value of the --db_res flag.\n",
      "  -acq, --acquired      Run resfinder for acquired resistance genes\n",
      "  -ao ACQ_OVERLAP, --acq_overlap ACQ_OVERLAP\n",
      "                        Genes are allowed to overlap this number of\n",
      "                        nucleotides. Default: 30.\n",
      "  -l MIN_COV, --min_cov MIN_COV\n",
      "                        Minimum (breadth-of) coverage of ResFinder within the\n",
      "                        range 0-1.\n",
      "  -t THRESHOLD, --threshold THRESHOLD\n",
      "                        Threshold for identity of ResFinder within the range\n",
      "                        0-1.\n",
      "  -d, --disinfectant    Run resfinder for disinfectant resistance genes\n",
      "  -db_disinf DB_PATH_DISINF, --db_path_disinf DB_PATH_DISINF\n",
      "                        Path to the databases for DisinFinder.\n",
      "  -db_disinf_kma DB_PATH_DISINF_KMA, --db_path_disinf_kma DB_PATH_DISINF_KMA\n",
      "                        Path to the DisinFinder databases indexed with KMA.\n",
      "                        Defaults to the value of the --db_res flag.\n",
      "  -c, --point           Run pointfinder for chromosomal mutations\n",
      "  -db_point DB_PATH_POINT, --db_path_point DB_PATH_POINT\n",
      "                        Path to the databases for PointFinder\n",
      "  -db_point_kma DB_PATH_POINT_KMA, --db_path_point_kma DB_PATH_POINT_KMA\n",
      "                        Path to the PointFinder databases indexed with KMA.\n",
      "                        Defaults to the value of the --db_path_point flag.\n",
      "  -g SPECIFIC_GENE [SPECIFIC_GENE ...], --specific_gene SPECIFIC_GENE [SPECIFIC_GENE ...]\n",
      "                        Specify genes existing in the database to search for -\n",
      "                        if none is specified all genes are included in the\n",
      "                        search.\n",
      "  -u, --unknown_mut     Show all mutations found even if in unknown to the\n",
      "                        resistance database\n",
      "  -l_p MIN_COV_POINT, --min_cov_point MIN_COV_POINT\n",
      "                        Minimum (breadth-of) coverage of Pointfinder within\n",
      "                        the range 0-1. If None is selected, the minimum\n",
      "                        coverage of ResFinder will be used.\n",
      "  -t_p THRESHOLD_POINT, --threshold_point THRESHOLD_POINT\n",
      "                        Threshold for identity of Pointfinder within the range\n",
      "                        0-1. If None is selected, the minimum coverage of\n",
      "                        ResFinder will be used.\n",
      "  --ignore_indels       Ignore frameshift-causing indels in Pointfinder.\n",
      "  --ignore_stop_codons  Ignore premature stop codons in Pointfinder.\n",
      "  -v, --version         Show program's version number and exit\n",
      "  --pickle              Create a pickle dump of the Isolate object. Currently\n",
      "                        needed in the CGE webserver. Dependency and this\n",
      "                        option is being removed.\n"
     ]
    }
   ],
   "source": [
    "!python -m resfinder --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: KMA did not run as expected.\n",
      "KMA finished with the following response:\n",
      "\n",
      "Filename:\t/home/npilquinao/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/resfinder_db/aminoglycoside.length.b\n",
      "Error: 2 (No such file or directory)\n",
      "\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\ncd /home/npilquinao/group_storage/npilquinao/ISP/2carbapenemase/samples/\\n\\npython -m resfinder\\\\\\n    --inputfastq\\\\\\n        VA61_2022/61_22_S82_R1_001.fastq.gz\\\\\\n        VA61_2022/61_22_S82_R2_001.fastq.gz\\\\\\n    --outputPath ~/BEM-ISP/ResFinder/VA61_2022/\\\\\\n    --acquired --disinfectant\\\\\\n    --db_path_res ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/resfinder_db/\\\\\\n    --db_path_disinf ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/disinfinder_db/\\\\\\n    --db_path_point ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/pointfinder_db/\\\\\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mcd /home/npilquinao/group_storage/npilquinao/ISP/2carbapenemase/samples/\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mpython -m resfinder\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --inputfastq\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        VA61_2022/61_22_S82_R1_001.fastq.gz\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        VA61_2022/61_22_S82_R2_001.fastq.gz\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --outputPath ~/BEM-ISP/ResFinder/VA61_2022/\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --acquired --disinfectant\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --db_path_res ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/resfinder_db/\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --db_path_disinf ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/disinfinder_db/\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    --db_path_point ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/pointfinder_db/\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/ISP/lib/python3.11/site-packages/IPython/core/interactiveshell.py:2541\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2539\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2540\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2541\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2543\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2544\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2545\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2546\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/miniconda3/envs/ISP/lib/python3.11/site-packages/IPython/core/magics/script.py:155\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    153\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    154\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 155\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/ISP/lib/python3.11/site-packages/IPython/core/magics/script.py:315\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    314\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 315\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\ncd /home/npilquinao/group_storage/npilquinao/ISP/2carbapenemase/samples/\\n\\npython -m resfinder\\\\\\n    --inputfastq\\\\\\n        VA61_2022/61_22_S82_R1_001.fastq.gz\\\\\\n        VA61_2022/61_22_S82_R2_001.fastq.gz\\\\\\n    --outputPath ~/BEM-ISP/ResFinder/VA61_2022/\\\\\\n    --acquired --disinfectant\\\\\\n    --db_path_res ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/resfinder_db/\\\\\\n    --db_path_disinf ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/disinfinder_db/\\\\\\n    --db_path_point ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/pointfinder_db/\\\\\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cd /home/npilquinao/group_storage/npilquinao/ISP/2carbapenemase/samples/\n",
    "\n",
    "python -m resfinder\\\n",
    "    --inputfastq\\\n",
    "        VA61_2022/61_22_S82_R1_001.fastq.gz\\\n",
    "        VA61_2022/61_22_S82_R2_001.fastq.gz\\\n",
    "    --outputPath ~/BEM-ISP/ResFinder/VA61_2022/\\\n",
    "    --acquired --disinfectant\\\n",
    "    --db_path_res ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/resfinder_db/\\\n",
    "    --db_path_disinf ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/disinfinder_db/\\\n",
    "    --db_path_point ~/group_storage/npilquinao/ISP/2carbapenemase/resfinder_databases/pointfinder_db/\\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Desorden__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import resfinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/npilquinao/miniconda3/envs/ISP/lib/python3.11/site-packages/resfinder/__init__.py\n"
     ]
    }
   ],
   "source": [
    "print(resfinder.__file__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "amr_abbreviations.md\t  __init__.py  run_resfinder.py\n",
      "cge\t\t\t  __main__.py  species_abbreviations.md\n",
      "environment_variables.md  __pycache__\n",
      "\n",
      "#!/usr/bin/env python3\n",
      "import io\n",
      "import sys\n",
      "import os\n",
      "import subprocess\n",
      "from argparse import ArgumentParser\n",
      "import pickle\n",
      "import json\n",
      "import hashlib\n",
      "\n",
      "from cgelib.output.result import Result\n",
      "from cgelib.utils.loaders_mixin import LoadersMixin\n",
      "from cgelib.utils.pliers_mixin import PliersMixin\n",
      "\n",
      "from resfinder.cge.config import Config\n",
      "from resfinder.cge.resfinder import ResFinder\n",
      "from resfinder.cge.pointfinder import PointFinder\n",
      "from resfinder.cge.output.std_results import ResFinderResultHandler\n",
      "from resfinder.cge.output.std_results import PointFinderResultHandler\n",
      "\n",
      "#  Modules used to create the extended ResFinder output (phenotype output)\n",
      "from resfinder.cge.phenotype2genotype.isolate import Isolate\n",
      "from resfinder.cge.phenotype2genotype.res_profile import PhenoDB\n",
      "from resfinder.cge.phenotype2genotype.res_sumtable import ResSumTable\n",
      "from resfinder.cge.phenotype2genotype.res_sumtable import PanelNameError\n",
      "\n",
      "from resfinder import __version__\n",
      "\n",
      "\n",
      "def eprint(*args, **kwargs):\n",
      "    print(*args, file=sys.stderr, **kwargs)\n",
      "\n",
      "\n",
      "def get_software_exec_res(conf: Config) -> dict:\n",
      "    software_exec_res = {\n",
      "        \"type\": \"software_exec\",\n",
      "        \"command\": \" \".join(sys.argv),\n",
      "        \"parameters\": get_call_parameters(conf)\n",
      "    }\n",
      "    software_exec_res[\"key\"] = hashlib.sha1(\n",
      "        bytes(software_exec_res[\"command\"], 'UTF-8')).hexdigest()\n",
      "    return software_exec_res\n",
      "\n",
      "\n",
      "def get_call_parameters(conf: Config) -> dict:\n",
      "    parameters = vars(conf).copy()\n",
      "    del (parameters['amr_abbreviations'])\n",
      "    return parameters\n",
      "\n",
      "\n",
      "def main():\n",
      "\n",
      "    # version = read_version(\"VERSION\")\n",
      "    version = __version__\n",
      "\n",
      "    ##########################################################################\n",
      "    # PARSE COMMAND LINE OPTIONS\n",
      "    ##########################################################################\n",
      "\n",
      "    parser = ArgumentParser(allow_abbrev=False)\n",
      "\n",
      "    # General options\n",
      "    parser.add_argument(\"-ifa\", \"--inputfasta\",\n",
      "                        help=\"Input fasta file.\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-ifq\", \"--inputfastq\",\n",
      "                        help=(\"Input fastq file(s). Assumed to be single-end \"\n",
      "                              \"fastq if only one file is provided, and assumed\"\n",
      "                              \" to be paired-end data if two files are \"\n",
      "                              \"provided.\"),\n",
      "                        nargs=\"+\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--nanopore\",\n",
      "                        action=\"store_true\",\n",
      "                        dest=\"nanopore\",\n",
      "                        help=\"If nanopore data is used\",\n",
      "                        default=False)\n",
      "    parser.add_argument(\"-o\", \"--outputPath\",\n",
      "                        help=(\"Output directory. If it doesn't exist, it will \"\n",
      "                              \"be created.\"),\n",
      "                        required=True,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-j\", \"--out_json\",\n",
      "                        help=(\"Specify JSON filename and output directory. If \"\n",
      "                              \"the directory doesn't exist, it will be \"\n",
      "                              \"created.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-b\", \"--blastPath\",\n",
      "                        help=\"Path to blastn\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-k\", \"--kmaPath\",\n",
      "                        help=\"Path to KMA\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-s\", \"--species\",\n",
      "                        help=\"Species in the sample\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--ignore_missing_species\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"If set, species is provided and --point flag \"\n",
      "                              \"is set, will not throw an error if no database \"\n",
      "                              \"is found for the provided species. If species \"\n",
      "                              \"is not found. Point mutations will silently \"\n",
      "                              \"be ignored.\"),\n",
      "                        default=False)\n",
      "    parser.add_argument(\"--output_aln\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"will add the alignments in the json output.\",\n",
      "                        default=False)\n",
      "\n",
      "    # Acquired resistance options\n",
      "    parser.add_argument(\"-db_res\", \"--db_path_res\",\n",
      "                        help=(\"Path to the databases for ResFinder.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_res_kma\", \"--db_path_res_kma\",\n",
      "                        help=(\"Path to the ResFinder databases indexed with \"\n",
      "                              \"KMA. Defaults to the value of the --db_res \"\n",
      "                              \"flag.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-acq\", \"--acquired\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Run resfinder for acquired resistance genes\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-ao\", \"--acq_overlap\",\n",
      "                        help=\"Genes are allowed to overlap this number of\\\n",
      "                              nucleotides. Default: {}.\".format(\n",
      "                            Config.DEFAULT_VALS[\"acq_overlap\"]),\n",
      "                        type=int,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-l\", \"--min_cov\",\n",
      "                        help=(\"Minimum (breadth-of) coverage of ResFinder \"\n",
      "                              \"within the range 0-1.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-t\", \"--threshold\",\n",
      "                        help=(\"Threshold for identity of ResFinder within the \"\n",
      "                              \"range 0-1.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    # Disinfectant resistance options\n",
      "    parser.add_argument(\"-d\", \"--disinfectant\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Run resfinder for disinfectant resistance genes\",\n",
      "                        default=False)\n",
      "    parser.add_argument(\"-db_disinf\", \"--db_path_disinf\",\n",
      "                        help=(\"Path to the databases for DisinFinder.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_disinf_kma\", \"--db_path_disinf_kma\",\n",
      "                        help=(\"Path to the DisinFinder databases indexed with \"\n",
      "                              \"KMA. Defaults to the value of the --db_res \"\n",
      "                              \"flag.\"),\n",
      "                        default=None)\n",
      "\n",
      "    # Point resistance option\n",
      "    parser.add_argument(\"-c\", \"--point\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Run pointfinder for chromosomal mutations\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_point\", \"--db_path_point\",\n",
      "                        help=\"Path to the databases for PointFinder\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-db_point_kma\", \"--db_path_point_kma\",\n",
      "                        help=(\"Path to the PointFinder databases indexed with \"\n",
      "                              \"KMA. Defaults to the value of the \"\n",
      "                              \"--db_path_point flag.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-g\", \"--specific_gene\",\n",
      "                        nargs='+',\n",
      "                        help=\"Specify genes existing in the database to \\\n",
      "                              search for - if none is specified all genes are \\\n",
      "                              included in the search.\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-u\", \"--unknown_mut\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"Show all mutations found even if in unknown to \"\n",
      "                              \"the resistance database\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-l_p\", \"--min_cov_point\",\n",
      "                        help=(\"Minimum (breadth-of) coverage of Pointfinder \"\n",
      "                              \"within the range 0-1. If None is selected, the \"\n",
      "                              \"minimum coverage of ResFinder will be used.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-t_p\", \"--threshold_point\",\n",
      "                        help=(\"Threshold for identity of Pointfinder within \"\n",
      "                              \"the range 0-1. If None is selected, the minimum\"\n",
      "                              \" coverage of ResFinder will be used.\"),\n",
      "                        type=float,\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--ignore_indels\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"Ignore frameshift-causing indels in \"\n",
      "                              \"Pointfinder.\"),\n",
      "                        default=None)\n",
      "    parser.add_argument(\"--ignore_stop_codons\",\n",
      "                        action=\"store_true\",\n",
      "                        help=\"Ignore premature stop codons in Pointfinder.\",\n",
      "                        default=None)\n",
      "    parser.add_argument(\"-v\", \"--version\", action=\"version\",\n",
      "                        version=__version__,\n",
      "                        help=\"Show program's version number and exit\")\n",
      "\n",
      "    # Temporary option only available temporary\n",
      "    parser.add_argument(\"--pickle\",\n",
      "                        action=\"store_true\",\n",
      "                        help=(\"Create a pickle dump of the Isolate object. \"\n",
      "                              \"Currently needed in the CGE webserver. \"\n",
      "                              \"Dependency and this option is being removed.\"),\n",
      "                        default=False)\n",
      "\n",
      "    args = parser.parse_args()\n",
      "\n",
      "    # Parse and check all arguments and expected files.\n",
      "    conf = Config(args)\n",
      "\n",
      "    # Initialise result dict\n",
      "    std_result = Result.init_software_result(\n",
      "        name=\"ResFinder\",\n",
      "        gitdir=f\"{conf.resfinder_root}/../../\")\n",
      "\n",
      "    init_result_data = {\n",
      "        \"provided_species\": conf.species,\n",
      "        \"software_version\": __version__,\n",
      "        \"key\": f\"ResFinder-{__version__}\",\n",
      "    }\n",
      "    std_result.add(**init_result_data)\n",
      "\n",
      "    if conf.acquired:\n",
      "        std_result.init_database(\"ResFinder\", conf.db_path_res)\n",
      "    if conf.point:\n",
      "        std_result.init_database(\"PointFinder\", conf.db_path_point_root)\n",
      "    if conf.disinf:\n",
      "        std_result.init_database(\"DisinFinder\", conf.db_path_disinf)\n",
      "\n",
      "    std_result.add_class(cl=\"software_executions\", **\n",
      "                         get_software_exec_res(conf))\n",
      "\n",
      "    # Load genotype to phenotype database\n",
      "    res_pheno_db = PhenoDB(\n",
      "        abclassdef_file=conf.abclassdef_file,\n",
      "        acquired_file=conf.phenotype_file,\n",
      "        point_file=conf.point_file,\n",
      "        disinf_file=conf.disinf_file,\n",
      "        disclassdef_file=conf.disclassdef_file\n",
      "    )\n",
      "\n",
      "    ##########################################################################\n",
      "    # ResFinder\n",
      "    ##########################################################################\n",
      "\n",
      "    if conf.acquired is True:\n",
      "\n",
      "        blast_results = None\n",
      "        kma_run = None\n",
      "\n",
      "        # Actually running ResFinder (for acquired resistance)\n",
      "        acquired_finder = ResFinder(db_conf_file=conf.db_config_file,\n",
      "                                    db_path=conf.db_path_res,\n",
      "                                    pheno_file=conf.phenotype_file,\n",
      "                                    notes=conf.db_notes_file,\n",
      "                                    db_path_kma=conf.db_path_res_kma)\n",
      "\n",
      "        if (conf.inputfasta):\n",
      "            blast_results = acquired_finder.blast(\n",
      "                inputfile=conf.inputfasta,\n",
      "                out_path=conf.outPath_res_blast,\n",
      "                min_cov=conf.rf_gene_cov,\n",
      "                threshold=conf.rf_gene_id,\n",
      "                blast=conf.blast,\n",
      "                allowed_overlap=conf.rf_overlap\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            acquired_finder.write_results(out_path=conf.outputPath,\n",
      "                                          result=blast_results,\n",
      "                                          res_type=ResFinder.TYPE_BLAST,\n",
      "                                          software=\"ResFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       blast_results,\n",
      "                                                       \"ResFinder\",\n",
      "                                                       conf)\n",
      "\n",
      "        else:\n",
      "            kma_nano_args = None\n",
      "            if conf.nanopore:\n",
      "                kma_nano_args = '-ont -md 5'\n",
      "\n",
      "            kma_run = acquired_finder.kma(\n",
      "                inputfile_1=conf.inputfastq_1,\n",
      "                inputfile_2=conf.inputfastq_2,\n",
      "                out_path=conf.outPath_res_kma,\n",
      "                db_path_kma=conf.db_path_res_kma,\n",
      "                min_cov=conf.rf_gene_cov,\n",
      "                threshold=conf.rf_gene_id,\n",
      "                kma_path=conf.kma,\n",
      "                databases=acquired_finder.databases,\n",
      "                sample_name=\"\",\n",
      "                kma_cge=True,\n",
      "                kma_apm=\"p\",\n",
      "                kma_1t1=True,\n",
      "                kma_add_args=kma_nano_args\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            acquired_finder.write_results(out_path=conf.outputPath,\n",
      "                                          result=kma_run.results,\n",
      "                                          res_type=ResFinder.TYPE_KMA,\n",
      "                                          software=\"ResFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       kma_run,\n",
      "                                                       \"ResFinder\",\n",
      "                                                       conf)\n",
      "    ##########################################################################\n",
      "    # DisinFinder\n",
      "    ##########################################################################\n",
      "    if (conf.disinf is True):\n",
      "\n",
      "        blast_results = None\n",
      "        kma_run = None\n",
      "\n",
      "        # Actually running DisinFinder (for disinfectant resistance)\n",
      "        disinf_finder = ResFinder(db_conf_file=conf.db_config_disinf_file,\n",
      "                                  db_path=conf.db_path_disinf,\n",
      "                                  pheno_file=conf.disinf_file,\n",
      "                                  notes=conf.db_notes_disinf_file,\n",
      "                                  db_path_kma=conf.db_path_disinf_kma)\n",
      "\n",
      "        if (conf.inputfasta):\n",
      "            blast_results = disinf_finder.blast(\n",
      "                inputfile=conf.inputfasta,\n",
      "                out_path=conf.outPath_disinf_blast,\n",
      "                min_cov=conf.dis_gene_cov,\n",
      "                threshold=conf.dis_gene_id,\n",
      "                blast=conf.blast,\n",
      "                allowed_overlap=conf.dis_overlap\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            disinf_finder.write_results(out_path=conf.outputPath,\n",
      "                                        result=blast_results,\n",
      "                                        res_type=ResFinder.TYPE_BLAST,\n",
      "                                        software=\"DisinFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       blast_results,\n",
      "                                                       \"DisinFinder\",\n",
      "                                                       conf)\n",
      "\n",
      "        else:\n",
      "            kma_nano_args = None\n",
      "            if conf.nanopore:\n",
      "                kma_nano_args = '-ont -md 5'\n",
      "\n",
      "            kma_run = disinf_finder.kma(\n",
      "                inputfile_1=conf.inputfastq_1,\n",
      "                inputfile_2=conf.inputfastq_2,\n",
      "                out_path=conf.outPath_disinf_kma,\n",
      "                db_path_kma=conf.db_path_disinf_kma,\n",
      "                min_cov=conf.dis_gene_cov,\n",
      "                threshold=conf.dis_gene_id,\n",
      "                kma_path=conf.kma,\n",
      "                databases=disinf_finder.databases,\n",
      "                sample_name=\"\",\n",
      "                kma_cge=True,\n",
      "                kma_apm=\"p\",\n",
      "                kma_1t1=True,\n",
      "                kma_add_args=kma_nano_args\n",
      "            )\n",
      "\n",
      "            # DEPRECATED\n",
      "            # TODO: make a write method that depends on the json output\n",
      "            disinf_finder.write_results(out_path=conf.outputPath,\n",
      "                                        result=kma_run.results,\n",
      "                                        res_type=ResFinder.TYPE_KMA,\n",
      "                                        software=\"DisinFinder\")\n",
      "\n",
      "            ResFinderResultHandler.standardize_results(std_result,\n",
      "                                                       kma_run,\n",
      "                                                       \"DisinFinder\",\n",
      "                                                       conf)\n",
      "    ##########################################################################\n",
      "    # PointFinder\n",
      "    ##########################################################################\n",
      "\n",
      "    if (conf.point):\n",
      "\n",
      "        blast_results = None\n",
      "        kma_run = None\n",
      "\n",
      "        finder = PointFinder(db_path=conf.db_path_point,\n",
      "                             species=conf.species_dir,\n",
      "                             gene_list=conf.specific_gene,\n",
      "                             ignore_indels=conf.ignore_indels,\n",
      "                             ignore_stop_codons=conf.ignore_stop_codons)\n",
      "\n",
      "        if (conf.inputfasta):\n",
      "\n",
      "            method = PointFinder.TYPE_BLAST\n",
      "\n",
      "            blast_run = finder.blast(inputfile=conf.inputfasta,\n",
      "                                     out_path=conf.outPath_point_blast,\n",
      "                                     min_cov=0.01,  # Sorts on coverage later\n",
      "                                     threshold=conf.pf_gene_id,\n",
      "                                     blast=conf.blast,\n",
      "                                     cut_off=False)\n",
      "            results = blast_run.results\n",
      "\n",
      "        else:\n",
      "            kma_nano_args = None\n",
      "            if conf.nanopore:\n",
      "                kma_nano_args = '-ont -md 5'\n",
      "            \n",
      "            method = PointFinder.TYPE_KMA\n",
      "            \n",
      "            kma_run = finder.kma(inputfile_1=conf.inputfastq_1,\n",
      "                                    inputfile_2=conf.inputfastq_2,\n",
      "                                    out_path=conf.outPath_point_kma,\n",
      "                                    db_path_kma=conf.db_path_point,\n",
      "                                    databases=[conf.species_dir],\n",
      "                                    min_cov=0.01,  # Sorts on coverage later\n",
      "                                    threshold=conf.pf_gene_id,\n",
      "                                    kma_path=conf.kma,\n",
      "                                    sample_name=conf.sample_name,\n",
      "                                    kma_cge=True,\n",
      "                                    kma_apm=\"p\",\n",
      "                                    kma_1t1=True,\n",
      "                                    kma_add_args=kma_nano_args)\n",
      "\n",
      "            results = kma_run.results\n",
      "        \n",
      "        PointFinderResultHandler.rename_promoter_results(results, method)\n",
      "        \n",
      "        if (conf.specific_gene):\n",
      "            results = PointFinder.discard_from_dict(\n",
      "                in_dict=results, wanted_list=conf.specific_gene)\n",
      "\n",
      "        if (method == PointFinder.TYPE_BLAST):\n",
      "            results_pnt = finder.find_best_seqs(results, conf.pf_gene_cov)\n",
      "        else:\n",
      "            results_pnt = results[finder.species]\n",
      "            if (results_pnt == \"No hit found\"):\n",
      "                results_pnt = {}\n",
      "            else:\n",
      "                results_pnt[\"excluded\"] = results[\"excluded\"]\n",
      "\n",
      "        # DEPRECATED\n",
      "        # mutations in raw reads is only found with write_results.\n",
      "        # TODO: make a write method that depends on the json output\n",
      "        finder.write_results(\n",
      "            out_path=conf.outputPath, result=results, res_type=method,\n",
      "            unknown_flag=conf.unknown_mut, min_cov=conf.pf_gene_cov,\n",
      "            perc_iden=conf.pf_gene_id)\n",
      "\n",
      "        if not conf.unknown_mut:\n",
      "            results_pnt = PointFinder.discard_unknown_muts(\n",
      "                results_pnt=results_pnt, phenodb=res_pheno_db, method=method)\n",
      "\n",
      "        PointFinderResultHandler.standardize_results(std_result,\n",
      "                                                     results_pnt,\n",
      "                                                     \"PointFinder\")\n",
      "\n",
      "    ##########################################################################\n",
      "    # Phenotype to genotype\n",
      "    ##########################################################################\n",
      "\n",
      "    # Isolate object store results\n",
      "    isolate = Isolate(name=conf.sample_name, species=conf.species,\n",
      "                      amr_panel_file=conf.db_panels_file)\n",
      "\n",
      "    if (conf.acquired or conf.disinf):\n",
      "        isolate.load_finder_results(std_table=std_result,\n",
      "                                    phenodb=res_pheno_db,\n",
      "                                    type=\"seq_regions\")\n",
      "    if (conf.point):\n",
      "        isolate.load_finder_results(std_table=std_result,\n",
      "                                    phenodb=res_pheno_db,\n",
      "                                    type=\"seq_variations\")\n",
      "\n",
      "    isolate.calc_res_profile(res_pheno_db)\n",
      "    ResFinderResultHandler.load_res_profile(std_result, isolate,\n",
      "                                            conf.amr_abbreviations)\n",
      "\n",
      "    if (conf.out_json):\n",
      "        std_result_file = conf.out_json\n",
      "    else:\n",
      "        std_result_file = \"{}/{}.json\".format(\n",
      "            conf.outputPath, conf.sample_name.replace(\"_R1\", \"\").split(\".\")[0])\n",
      "    with open(std_result_file, 'w') as fh:\n",
      "        fh.write(std_result.json_dumps())\n",
      "\n",
      "    # Create and write the downloadable tab file\n",
      "    pheno_profile_str = isolate.profile_to_str_table(header=True)\n",
      "\n",
      "    # TODO: REMOVE THE NEED FOR THE PICKLED FILE\n",
      "    if (conf.pickle):\n",
      "        isolate_pickle = open(\"{}/isolate.p\".format(conf.outputPath), \"wb\")\n",
      "        pickle.dump(isolate, isolate_pickle, protocol=2)\n",
      "\n",
      "    pheno_table_file = \"{}/pheno_table.txt\".format(conf.outputPath)\n",
      "    with open(pheno_table_file, 'w') as fh:\n",
      "        fh.write(pheno_profile_str)\n",
      "\n",
      "    if (conf.species is not None):\n",
      "        # Apply AMR panel\n",
      "        input_amr_panels = \"{}/phenotype_panels.txt\".format(conf.db_path_res)\n",
      "        res_sum_table = ResSumTable(pheno_profile_str)\n",
      "        res_sum_table.load_amr_panels(input_amr_panels)\n",
      "\n",
      "        try:\n",
      "            panel_profile_str = res_sum_table.get_amr_panel_str(\n",
      "                panel_name_raw=conf.species, header=True)\n",
      "        # If specified species does not have an associated panel, just ignore it\n",
      "        # and exit.\n",
      "        except PanelNameError:\n",
      "            if conf.species != 'other':\n",
      "                eprint(\"Warning: No panel was detected for the species: {}\"\n",
      "                       .format(conf.species))\n",
      "            sys.exit()\n",
      "\n",
      "        amr_panel_filename = conf.species.replace(\" \", \"_\")\n",
      "\n",
      "        panel_tabel_file = (\"{tablename}_{species}.txt\"\n",
      "                            .format(tablename=pheno_table_file[:-4],\n",
      "                                    species=amr_panel_filename))\n",
      "\n",
      "        with open(panel_tabel_file, \"w\") as fh:\n",
      "            fh.write(panel_profile_str)\n",
      "\n",
      "    return 0\n",
      "\n",
      "\n",
      "if __name__ == '__main__':\n",
      "    sys.exit(main())\n"
     ]
    }
   ],
   "source": [
    "!ls /home/npilquinao/miniconda3/envs/ISP/lib/python3.11/site-packages/resfinder/\n",
    "print(\"\")\n",
    "!cat /home/npilquinao/miniconda3/envs/ISP/lib/python3.11/site-packages/resfinder/run_resfinder.py"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ISP",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
